inst/include/dust/dust.hpp

```diff
@@ -0,0 +1,140 @@
+#ifndef DUST_DUST_HPP
+#define DUST_DUST_HPP
+
+#include <dust/rng.hpp>
+
+#include <utility>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace dust {
+
+template <typename T>
+class Particle {
+public:
+  typedef typename T::init_t init_t;
+  Particle(init_t data, size_t step) :
+    _model(data),
+    _step(step),
+    _y(_model.initial(_step)),
+    _y_swap(_model.size()) {
+  }
+
+  void run(const size_t step_end, RNG& rng) {
+    while (_step < step_end) {
+      _model.update(_step, _y, rng, _y_swap);
+      _step++;
+      std::swap(_y, _y_swap);
+    }
+  }
+
+  void state(const std::vector<size_t>& index_y,
+             std::vector<double>::iterator end_state) const {
+    for (size_t i = 0; i < index_y.size(); ++i) {
+      *(end_state + i) = _y[index_y[i]];
+    }
+  }
+
+  void state(std::vector<double>::iterator end_state) const {
+    for (size_t i = 0; i < _y.size(); ++i) {
+      *(end_state + i) = _y[i];
+    }
+  }
+
+  size_t size() const {
+    return _y.size();
+  }
+
+  size_t step() const {
+    return _step;
+  }
+
+private:
+  T _model;
+  size_t _step;
+
+  std::vector<double> _y;
+  std::vector<double> _y_swap;
+};
+
+template <typename T>
+class Dust {
+public:
+  typedef typename T::init_t init_t;
+  Dust(init_t data, size_t step, const std::vector<size_t> index_y,
+       const size_t n_particles, const size_t n_threads,
+       const size_t n_generators, const double seed) :
+    _index_y(index_y),
+    _n_threads(n_threads),
+    _rng(n_generators, seed) {
+    for (size_t i = 0; i < n_particles; ++i) {
+      _particles.push_back(Particle<T>(data, step));
+    }
+  }
+
+  void reset(init_t data, size_t step) {
+    size_t n_particles = _particles.size();
+    _particles.clear();
+    for (size_t i = 0; i < n_particles; ++i) {
+      _particles.push_back(Particle<T>(data, step));
+    }
+  }
+
+  // This scheme means that if we have the same number of generators
+  // and threads then work for thread i occurs on generator i.
+  //
+  // If we have more generators than threads then each thread uses m =
+  // n_generators / n_threads (so if we had 8 generators and 2
+  // threads, each thread would use 4 generators).  The first thread
+  // will cycle alternate through the first m generators, the second
+  // thread will cycle through the second m generators, etc.
```

Comment:
To see how this might work in practice, see:

```r
n_generators <- 8
n_threads <- 2
m <- n_generators / n_threads
i <- 0:24

thread_idx <- 0
i %% m + thread_idx * m

thread_idx <- 1
i %% m + thread_idx * m

# ----

n_generators <- 8
n_threads <- 4
m <- n_generators / n_threads
i <- 0:24

thread_idx <- 0
i %% m + thread_idx * m

thread_idx <- 1
i %% m + thread_idx * m

thread_idx <- 2
i %% m + thread_idx * m

thread_idx <- 3
i %% m + thread_idx * m
```